Require AWS SDK v3

stefanwork opened the pull request:
I am trying to fix the flagship so that it isn't bundling all 222 AWS gems, and it looks like this gem is causing it.

@gaorlov Do you think we can release a new version of this? I'm not quite sure how to push a new version, but I can learn how to do it if you think it's Ok? I'll then create the flagship PR to upgrade this gem. Thanks!

stefanwork commented:
Looks like we need to fix Circle's access to GitHub.. 🤔

stefanwork commented:
Here's the resulting `Gemfile.lock`:

```
PATH
  remote: .
  specs:
    conrad (2.4.0)
      activemodel
      aws-sdk-kinesis
      aws-sdk-sqs

GEM
  remote: https://rubygems.org/
  specs:
    activemodel (6.0.2.2)
      activesupport (= 6.0.2.2)
    activesupport (6.0.2.2)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
      zeitwerk (~> 2.2)
    ast (2.4.0)
    aws-eventstream (1.1.0)
    aws-partitions (1.304.0)
    aws-sdk-core (3.94.0)
      aws-eventstream (~> 1, >= 1.0.2)
      aws-partitions (~> 1, >= 1.239.0)
      aws-sigv4 (~> 1.1)
      jmespath (~> 1.0)
    aws-sdk-kinesis (1.21.0)
      aws-sdk-core (~> 3, >= 3.71.0)
      aws-sigv4 (~> 1.1)
    aws-sdk-sqs (1.24.0)
      aws-sdk-core (~> 3, >= 3.71.0)
      aws-sigv4 (~> 1.1)
    aws-sigv4 (1.1.3)
      aws-eventstream (~> 1.0, >= 1.0.2)
    concurrent-ruby (1.1.6)
    i18n (1.8.2)
      concurrent-ruby (~> 1.0)
    jaro_winkler (1.5.4)
    jmespath (1.4.0)
    minitest (5.14.0)
    minitest-stub-const (0.6)
    parallel (1.19.1)
    parser (2.7.1.1)
      ast (~> 2.4.0)
    powerpack (0.1.2)
    rainbow (3.0.0)
    rake (13.0.1)
    rubocop (0.60.0)
      jaro_winkler (~> 1.5.1)
      parallel (~> 1.10)
      parser (>= 2.5, != 2.5.1.1)
      powerpack (~> 0.1)
      rainbow (>= 2.2.2, < 4.0)
      ruby-progressbar (~> 1.7)
      unicode-display_width (~> 1.4.0)
    ruby-progressbar (1.10.1)
    thread_safe (0.3.6)
    tzinfo (1.2.7)
      thread_safe (~> 0.1)
    unicode-display_width (1.4.1)
    yard (0.9.24)
    zeitwerk (2.3.0)

PLATFORMS
  ruby

DEPENDENCIES
  conrad!
  minitest (~> 5.0)
  minitest-stub-const
  rake (>= 13.0)
  rubocop (~> 0.60.0)
  yard

BUNDLED WITH
   1.17.2
```

stefanwork commented:
By the way, it seems like 2.5.0 was not actually released? At least `VERSION` wasn't updated.

stefanwork commented:
It looks like this gem is currently compatible with both AWS SDK v2 and v3. Accounts is using v2.4.0 of this gem and AWS SDK v2. In the flagship we are also running v2.4.0 of this gem, and have been using AWS SDK v3 since https://github.com/getoutreach/outreach/pull/13560.

I could add the following dependency to force the use of AWS SDK v3:
```ruby
spec.add_dependency 'aws-sdk-core', '~> 3.0'
```

gaorlov commented:
> It looks like this gem is currently compatible with both AWS SDK v2 and v3. Accounts is using v2.4.0 of this gem and AWS SDK v2. In the flagship we are also running v2.4.0 of this gem, and have been using AWS SDK v3 since [getoutreach/outreach#13560](https://github.com/getoutreach/outreach/pull/13560).
> 
> I could add the following dependency to force the use of AWS SDK v3:
> 
> ```ruby
> spec.add_dependency 'aws-sdk-core', '~> 3.0'
> ```
> 
> EDIT: Actually, this is actually not necessary since `aws-sdk-core ~> 3, >= 3.71.0` is already a transient dependency in both `aws-sdk-sqs` and `aws-sdk-kinesis`. So this PR will already force AWS SDK v3 to be used without the line above. But adding the line may make it more explicit and prevent the use of a potential future AWS SDK v4.

we could also do something like 
```ruby
spec.add_dependency 'aws-sdk-core', '>= 3.0'
```

So long as you're comfortable forcing Accounts to use an older version of Conrad (until you can upgrade Accounts to use `aws-sdk` 3), LGTM. Let's be sure to bump the major version. Nice work! Thanks for taking the time to keep this updated!

stefanwork commented:
> So long as you're comfortable forcing Accounts to use an older version of Conrad (until you can upgrade Accounts to use `aws-sdk` 3), LGTM.

Accounts is actually not using the AWS SDK directly, it is just for this gem, so not a problem there. It should be fine to upgrade to v3.

Are you aware of any other project that is using this gem? I can't find any other project than the flagship and accounts. https://github.com/search?p=2&q=%40getoutreach+-%40getoutreach%2Fconrad+conrad&type=Code

> Let's be sure to bump the major version.

Sure. I can do it in this PR. :+1:

gaorlov commented:
In other projects i've worked on the version gets bumped in master.  @jpanderson-outreach what's the versioning procedure for this gem?

jpanderson-outreach commented:
Seems fine to me. This was only so "loosey goosey" before because the flagship and accounts were both on v2 at the time but we didn't want to force things on v3 to have to use v2 (and thus be unable to use this). 

I'm unsure why the builds are failing on the Checkout step since this is a public repo :/ 

as for versioning, historically I've done a separate PR that bumps the version on everything. Greg and I have push permissions for this gem in Rubygems.

stefanwork commented:
After kicking CircleCI a bit and readding the SSH key it uses, the tests are now working. 🎉 

I rebased this PR and cleaned up the commits. Will bump the version in a separate PR.
